File: Python/65-矩阵中的路径.py
# -*- coding:utf-8 -*-
class Solution:
    def hasPath(self, matrix, rows, cols, path):
        # write code here
        visit = [0] * (rows * cols)
        def helper(i, j, k):
            idx = i * cols + j
            if i < 0 or i >= rows or j < 0 or j >= cols or visit[idx] or matrix[idx] != path[k]:
                return False
            if k == len(path) - 1:
                return True
            visit[idx] = True
            if helper(i - 1, j, k + 1):
                return True
            if helper(i + 1, j, k + 1):
                return True
            if helper(i, j - 1, k + 1):
                return True
            if helper(i, j + 1, k + 1):
                return True
            visit[idx] = False
            return False
        for i in range(rows):
            for j in range(cols):
                if helper(i, j, 0):
                    return True
        return False

File: C++/65-矩阵中的路径.cpp
class Solution
{
public:
    bool hasPath(char *matrix, int rows, int cols, char *str)
    {
        bool visit[rows * cols];
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                visit[i * cols + j] = false;
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                if (helper(matrix, rows, cols, str, i, j, 0, visit))
                    return true;
        return false;
    }

    bool helper(char *mat, int &r, int &c, char *s, int i, int j, int k, bool visit[])
    {
        int idx = i * c + j;
        if (i < 0 || j < 0 || i >= r || j >= c || visit[idx] || mat[idx] != s[k])
            return false;
        if (k == strlen(s) - 1)
            return true;
        visit[idx] = true;
        if (helper(mat, r, c, s, i - 1, j, k + 1, visit))
            return true;
        if (helper(mat, r, c, s, i + 1, j, k + 1, visit))
            return true;
        if (helper(mat, r, c, s, i, j - 1, k + 1, visit))
            return true;
        if (helper(mat, r, c, s, i, j + 1, k + 1, visit))
            return true;
        visit[idx] = false;
        return false;
    }
};